### PR TITLE
#164198181 Validates signup parameters

### PR DESCRIPTION
**Pull request by valenstical**
**What does this PR do?**
This PR validates registration/signup parameters before sending control over to the user controller.

**Description of Task to be completed**

- [x] Install express-validator to handle validation
- [x] Create middleware to validate each parameter
- [x] Create middleware to handle validation errors if any
- [x] Mount middleware created above to the user signup route
- [x] Write test against implementation
- [x] Create helper function to send response data
- [x] Create constant file to hold all re-usable constants

#### How should this be manually tested

_Prerequisite: Postgres_

1. Run the following commands on terminal:

```bash
git clone -b feature/164198181/validate-signup --single-branch \
https://github.com/andela/apollo-ah-backend.git

Navigate to the apollo-ah-backend directory
Run cp .env.example .env
npm install
```

2. Modify `.env` file with your db credentials

3. Run migrations with `npm run migrate`

4. Spin up the server by running `npm run start:dev` on your terminal

4. Make a POST request to `localhost:3000/api/v1/users` respectively with [Postman](https://www.getpostman.com/downloads/).

❗️ **Important:** Provide user payload when making request, sample below:
```json
{
  "email": "example@email.com",
  "password": "secret"
  "username" "user"
}
```

**NOTE:** To run test locally, create `.env.test` file. Run command: `cp .env.example .env.test`

#### Any background context you want to provide

N/A

#### What are the relevant pivotal tracker stories

[#164198181](https://www.pivotaltracker.com/story/show/164198181)

#### Screenshots
N/A

_-- Template format by @devdbrandy_

### CODE REVIEW

File: routes/users.js [hunk 1]
```diff
@@ -40,7 +42,10 @@ const router = express.Router();
  *         schema:
  *           $ref: '#/definitions/User'
  */
-router.post('/', UsersController.register);
+router.post('/',
```
[the remark this answers is not in the document]
Comment (valenstical) [reply routes/users.js]: Code adjusted.

File: routes/users.js [hunk 2]
```diff
@@ -40,7 +42,10 @@ const router = express.Router();
...
+  Validator.validateRegistration(),
+  handleRegistration,
+  UsersController.register);
```
[the remark this answers is not in the document]
Comment (valenstical) [reply routes/users.js]: Issue resolved.

File: controllers/users.js [hunk 3]
```diff
@@ -27,15 +29,31 @@ class UsersController {
...
+ * @param {*} value The actual value to test for
+ * @returns {object} The user object
+ * @memberof UsersController
+ */
+  static async getUser(param, value) {
+    let user;
+    try {
+      user = await User.findOne({ where: { [param]: value } });
+    } catch (error) {
+      logger.log(error);
```
[the remark this answers is not in the document]
Comment (valenstical) [reply controllers/users.js]: Issue resolved.

File: helpers/response.helper.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { STATUS } from './constants';
```
Comment (aanchirinah) [inline helpers/response.helper.js]: #### Critical
- The naming convention for this file is against BEM

Comment (valenstical) [reply helpers/response.helper.js]: Thank you @aanchirinah for pointing that out. I have renamed the file to the prescribed convention.

File: helpers/constants.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  USERNAME_EXITS: 'The username is already taken, try another one',
+};
+
+export const FIELD = {
+  USERNAME: 'username',
+  PASSWORD: 'password',
+  EMAIL: 'email',
+};
+
+export const expressFormater = ({
```
Comment (devdbrandy) [inline helpers/constants.js]: Please what does this function do? Would be nice if you could document it.

Comment (valenstical) [reply helpers/constants.js]: This function formats the object keys from express-validator to use a more generic naming pattern.

Comment (valenstical) [reply helpers/constants.js]: Proper documentation has been added. Thank you for pointing that out.

**Comment by devdbrandy** [comment]
Looks good to me.

**Comment by aanchirinah** [comment]
LGTM

**Pull request merged**